### PR TITLE
Add check for invalid r/s values on ecrecover

### PR DESCRIPTION
**Pull request by axic**
Following up #542. This implements the rest of the checks:
Screenshot

### CODE REVIEW

File: warplib/ecrecover.cairo [hunk 1]
```diff
@@ -26,6 +29,20 @@ func ecrecover_eth{range_check_ptr, bitwise_ptr : BitwiseBuiltin*, keccak_ptr :
 
     let v = v - 27
 
+    let (r_eq_0) = warp_eq256(_r, Uint256(0, 0))
+    let (s_eq_0) = warp_eq256(_s, Uint256(0, 0))
+
+    if r_eq_0 + s_eq_0 != 0:
```
Comment (axic) [inline warplib/ecrecover.cairo:35]: Neither r/s or can be zero

File: warplib/ecrecover.cairo [hunk 2]
```diff
@@ -26,6 +29,20 @@ func ecrecover_eth{range_check_ptr, bitwise_ptr : BitwiseBuiltin*, keccak_ptr :
...
+        return (0)
+    end
+
+    let (r_lt_n) = warp_lt256(_r, SECP256K1_N)
+    let (s_lt_n) = warp_lt256(_s, SECP256K1_N)
+
+    if r_lt_n + s_lt_n != 2:
```
Comment (axic) [inline warplib/ecrecover.cairo:42]: r/s can't be over n

File: tests/behaviour/contracts/ecrecover/failing_ecrecover_invalid_input.sol [hunk 3]
```diff
@@ -1,12 +1,38 @@
 contract WARP {
+    uint256 constant SECP256K1_N = 115792089237316195423570985008687907852837564279074904382605163141518161494337;
+
     // ecrecover should return zero for malformed input
-	// (v should be 27 or 28, not 1)
-	// Note that the precompile does not return zero but returns nothing.
-    function f() public returns (uint160) {
+    // (v should be 27 or 28, not 1)
+    // Note that the precompile does not return zero but returns nothing.
+    function f() public returns (address) {
```
[the remark this answers is not in the document]
Comment (axic) [reply tests/behaviour/contracts/ecrecover/failing_ecrecover_invalid_input.sol:7]: Oh ok, was confused by that. This should be documented somewhere as now it is incompatible with typing rules of contracts, i.e. one cannot without modifications use warp with regular sources.

[the remark this answers is not in the document]
Comment (axic) [reply tests/behaviour/contracts/ecrecover/failing_ecrecover_invalid_input.sol:7]: @swapnilraj fixed it.

[the remark this answers is not in the document]
Comment (swapnilraj) [reply tests/behaviour/contracts/ecrecover/failing_ecrecover_invalid_input.sol]: Yea you're right, we'll include it in our public messaging for warp's release and add a note in the readme

File: warplib/ecrecover.cairo [hunk 4]
```diff
@@ -26,6 +29,20 @@ func ecrecover_eth{range_check_ptr, bitwise_ptr : BitwiseBuiltin*, keccak_ptr :
 
     let v = v - 27
 
+    let (r_eq_0) = warp_eq256(_r, Uint256(0, 0))
+    let (s_eq_0) = warp_eq256(_s, Uint256(0, 0))
```
Comment (swapnilraj) [inline warplib/ecrecover.cairo:33]: This file is missing import for warp_eq256

Comment (axic) [reply warplib/ecrecover.cairo:33]: Fixed.

File: warplib/ecrecover.cairo [hunk 5]
```diff
@@ -26,6 +29,20 @@ func ecrecover_eth{range_check_ptr, bitwise_ptr : BitwiseBuiltin*, keccak_ptr :
 
     let v = v - 27
 
+    let (r_eq_0) = warp_eq256(_r, Uint256(0, 0))
```
Comment (swapnilraj) [inline warplib/ecrecover.cairo:32]: `_r` is a typo, it should be `r_` same for _s
feel free to change the argument identifiers instead

Comment (axic) [reply warplib/ecrecover.cairo]: Yeah fixed it already, also added comments.

File: warplib/ecrecover.cairo [hunk 6]
```diff
@@ -20,12 +23,29 @@ func ecrecover_eth{range_check_ptr, bitwise_ptr : BitwiseBuiltin*, keccak_ptr :
     let (v_is_27) = warp_eq(v, 27)
     let (v_is_28) = warp_eq(v, 28)
 
+    # Expect one of them to be 1 (true)
     if v_is_27 + v_is_28 == 0:
```
Comment (axic) [inline warplib/ecrecover.cairo:27]: Should this be changed to `!= 1` for readability? Makes no difference though.

File: warplib/ecrecover.cairo [hunk 7]
```diff
@@ -10,7 +10,10 @@ from starkware.cairo.common.uint256 import Uint256
 from starkware.cairo.common.cairo_secp.bigint import BigInt3, uint256_to_bigint
 from starkware.cairo.common.cairo_secp.ec import EcPoint
 
-from warplib.maths.eq import warp_eq
+from warplib.maths.eq import warp_eq, warp_eq256
+from warplib.maths.lt import warp_lt256
+
+const SECP256K1_N = 115792089237316195423570985008687907852837564279074904382605163141518161494337
```
Comment (axic) [inline warplib/ecrecover.cairo:16]: Hm, this is actually close to 256 bits so it won't fit felt. Too bad there's no compiler warning for truncation.

File: warplib/ecrecover.cairo (hunk 6 above)
Comment (swapnilraj) [reply warplib/ecrecover.cairo:30]: In the context on of the early return on the next line, I think checking for 0 is more readable but I see your point as well. I think with the comment you added either of them are fine.

**Comment by axic** [comment]
@swapnilraj apologies for killing your CI with fixing things based on it, had issues running the test suite locally.

**Comment by swapnilraj** [comment]
> @swapnilraj apologies for killing your CI with fixing things based on it, had issues running the test suite locally.

No worries, keep hammering it as much as you want!

**Comment by axic** [comment]
@swapnilraj passed! 😅

**Comment by swapnilraj** [comment]
> @swapnilraj passed! 😅

Yep it all looks good to me, if you could apply this [patch](https://www.toptal.com/developers/hastebin/raw/bupiyazugi), since our behaviour tests do not make use of the solidity DSL for testing so we have to setup the expectations by hand. We only run our solidity DSL parser on the semantic tests from the solidity repo.

I ran the tests locally and they all pass with the patch as well, so once the patch is applied we can merge.

**Comment by axic** [comment]
> I ran the tests locally and they all pass with the patch as well, so once the patch is applied we can merge.

Good catch, added. Also the enable-maintainers-to-commit thing should be enabled for you to push to my branches, should you wish.

**Comment by axic** [comment]
Thanks!

**Pull request merged**